decoder/expression_candidates.go

```diff
@@ -152,15 +165,103 @@ func constraintsAtPos(expr hcl.Expression, constraints ExprConstraints, pos hcl.
 	return ExprConstraints{}, expr.Range()
 }
 
-func (d *PathDecoder) expressionCandidatesAtPos(constraints ExprConstraints, outerBodyRng, prefixRng, editRng hcl.Range) (lang.Candidates, error) {
-	candidates := lang.NewCandidates()
+type pathKey struct{}
+
+func PathFromContext(ctx context.Context) (lang.Path, bool) {
+	p, ok := ctx.Value(pathKey{}).(lang.Path)
+	return p, ok
+}
+
+type posKey struct{}
+
+func PosFromContext(ctx context.Context) (hcl.Pos, bool) {
+	p, ok := ctx.Value(posKey{}).(hcl.Pos)
+	return p, ok
+}
+
+type filenameKey struct{}
+
+func FilenameFromContext(ctx context.Context) (string, bool) {
+	f, ok := ctx.Value(filenameKey{}).(string)
+	return f, ok
+}
 
-	for _, c := range constraints {
-		candidates.List = append(candidates.List, d.constraintToCandidates(c, outerBodyRng, prefixRng, editRng)...)
+func isEmptyExpr(expr hclsyntax.Expression) bool {
+	l, ok := expr.(*hclsyntax.LiteralValueExpr)
+	if !ok {
+		return false
 	}
+	if l.Val != cty.DynamicVal {
+		return false
+	}
+	// TODO? more checks
 
-	candidates.IsComplete = true
-	return candidates, nil
+	return true
+}
+
+func isMultilineTemplateExpr(expr hclsyntax.Expression) bool {
+	t, ok := expr.(*hclsyntax.TemplateExpr)
+	if !ok {
+		return false
+	}
+	return t.Range().Start.Line != t.Range().End.Line
+}
+
+func (d *PathDecoder) candidatesFromHooks(ctx context.Context, attr *hclsyntax.Attribute, schema *schema.AttributeSchema, outerBodyRng hcl.Range, pos hcl.Pos) []lang.Candidate {
+	candidates := make([]lang.Candidate, 0)
+	expr := ExprConstraints(schema.Expr)
+	exprType, ok := expr.LiteralType()
+	if !ok && exprType != cty.String {
+		// Return early as we only support string values for now
+		return candidates
+	}
+
+	editRng := attr.Expr.Range()
+	if isEmptyExpr(attr.Expr) || isMultilineTemplateExpr(attr.Expr) {
+		// An empty expression or a string without a closing quote will lead to
+		// an attribute expression spanning multiple lines.
+		// Since text edits only support a single line, we're resetting the End
+		// position here.
+		editRng.End = pos
+	}
+	prefixRng := attr.Expr.Range()
+	prefixRng.End = pos
+	prefixBytes, _ := d.bytesFromRange(prefixRng)
+	prefix := string(prefixBytes)
+	prefix = strings.TrimLeft(prefix, `"`)
+
+	ctx = context.WithValue(ctx, pathKey{}, d.path)
+	ctx = context.WithValue(ctx, filenameKey{}, attr.Expr.Range().Filename)
+	ctx = context.WithValue(ctx, posKey{}, pos)
+
+	for _, hook := range schema.CompletionHooks {
+		if completionFunc, ok := d.decoderCtx.CompletionHooks[hook.Name]; ok {
+			res, _ := completionFunc(ctx, cty.StringVal(prefix))
+
+			for _, c := range res {
+				// We're adding quotes to the string here, as we're always
+				// replacing the whole edit range
+				text := fmt.Sprintf("%q", c.RawInsertText)
+
+				candidates = append(candidates, lang.Candidate{
+					Label:        text,
```

Comment:
We're using `text` instead of `c.Label` here, because VS Code is using the label value to filter the text edits. So they need to match. I could look into setting `filterText` on `lang.Candidate` instead 🤔

Reply:
I think this quoting logic should live in the helper which converts `decoder.ExpressionCandidate` to `decoder.Candidate` as this would not be relevant in any context other than if we're completing (string) expressions. I know there's currently no way we would complete anything, but it would be good to be prepared for that eventuality.

So `Label` can then stay set to `c.Label` here. The same goes for `RawInsertText` - I think the main reason we picked that name was to convey that it gets inserted "as-is" without side effects, so we should move the quoting of the text edit also.